### DISCUSSION TITLE
Support for symbolic links

### DISCUSSION THREAD

**Discussion by B4rc1** [srid/emanote#253] [Ideas]
**Describe the bug**
Emanote does not index or show symbolic links.

**To Reproduce**
1. Go to your notes directory
2. create a symbolic link from somewhere into your notes directory: e.g. `ln -s ~/uni/informatik-obsidian/Lecture\ Notes/ ./lecture-notes/`
3. Open Emanote
4. See no entry `lecture-notes`

**Expected behavior**
`lecture-notes` should be included like any other directory.

**Additional Context**
`ls -la` of my Notes directory (cut for brevity):
```
❯ ls -la
# ...
lrwxrwxrwx  50 jonas users 11 Feb 20:43  lecture-notes -> /home/jonas/uni/informatik-obsidian/Lecture Notes
# ...
```

**Comment by srid** [comment]
Symlinks are not supported in Emanote for good reasons. I don't see a strong reason to support it. If you had `ln -s / ./root`, should Emanote traverse all files in your filesystem?

**Reply by B4rc1** [reply]
That is what I would have expected, yes.

**Reply by B4rc1** [reply]
but yeah with layers there is no real point in supporting symlinks I agree.
It should be documented that symlinks are not supported and that layers are the intended alternative in order to avoid confusion.

**Reply by B4rc1** [reply]
after restarting emanote once, symlinks are working as expected, this is a bug right?

**Reply by B4rc1** [reply]
yeah so there is definitely something funky going on because even after deleting the symlink I can still acess the files the symlink pointed to like I never deleted it until I restart emanote

**Reply by srid** [reply]
What sequence of commands will reproduce this?

Are you using the run or gen command?

**Reply by B4rc1** [reply]
- `mkdir ~/my-cool-dir`
- `touch ~/my-cool-dir/test.md`
- `ln -s ~/my-cool-dir $NOTES_DIR`
- `pkill emanote` (if running)
- `emanote --layers=. run`

the `test` file is now aceesable in emanote in `my-cool-dir`

now issuing `rm $NOTES_DIR/my-cool-dir` removes the symlink from disk but emanote still displays it in the web view. The file can still be acessed.
After just trying to modify the original file (`~/my-cool-dir/test.md`) after I deleted the symlink I get this in the directory view:
![image](https://user-images.githubusercontent.com/10383737/153686860-aa613543-ee5e-41cc-afbc-23711d049f35.png)

`pkill emanote` now the directory is gone

**Reply by srid** [reply]
What's NOTES_DIR?

**Reply by B4rc1** [reply]
A placeholder for your notes dir (where you start emanote from)

**Reply by srid** [reply]
What is it set to? What is its value?

**Reply by B4rc1** [reply]
in my case `~/zk/`

**Reply by srid** [reply]
To be absolute clear, try using absolute paths everywhere. For example, instead of:

```
emanote --layers=. run
```

run:

```
emanote --layers=~/zk run
```

I cannot use your sequence of your commands to reproduce this, because I do not know what your current directory is.

**Reply by B4rc1** [reply]
I cannot use abosulte paths in emanote --layers. When executing `emanote --layers=~/zk/ run` I get `emanote: /home/jonas/~/zk/: getDirectoryContents:openDirStream: does not exist (No such file or directory)` which is obviously true since that file does not exist.

**Reply by B4rc1** [reply]
using `emanote --layers=/home/jonas/zk run` does work tho

**Reply by B4rc1** [reply]
the exact commands I ran taken from my shell history:
```
mkdir ~/my-cool-dir/
touch ~/my-cool-dir/test.md
ln -s ~/my-cool-dir ~/zk/
pkill emanote
emanote --layers=/home/jonas/zk/ run
```

using
```
❯ emanote --version
0.4.10.2
```

**Reply by B4rc1** [reply]
to be honest after playing around with working symlinks a bit I think it could still be valuable to keep (optional) support for them. The problem you mentioned with traversing the root directory also happens when I include the root directory as a layer.

The main benefit of keeping them is interaction with external tools, especially CLI since I don't have to remember where what layer or file is, instead its managed by my operating system and universally supported.

Lets maybe do an example: I want use fzf to choose a note to edit. With symlinks I can link my external directory into my main note directory and issue `nvim $(find -L ~/zk/ -type f | fzf)` (`find -L` means follow symbolic links) in order to pick the file I want to edit with fzf and then open it in neovim.
With layers there is no universal way to do that except remembering (or saving in a variable) every layer directory and then concateniting the results before picking them with fzf.

**Comment by srid** [comment]
You could try using layers, though:

```
emanote --layers=.;~/uni/informatik-obsidian/Lecture\ Notes/ run
```

This will merge the two folders.

**Reply by B4rc1** [reply]
Is there a way to specify the directory of the layer (e.g. in subdirectory`lecture-notes`)? I don't want `Lecture Notes` to be included in my root emanotes view since there are lots of loose files and it clutters up the directory view.

**Reply by B4rc1** [reply]
I cannot include the parent directory since that has a bunch of unrelated files as well.

**Reply by srid** [reply]
That's good idea: https://github.com/srid/emanote/issues/254
